### ISSUE TITLE
internal/eval: close()-d embedded disjunction of structs evaluated too early?

### ISSUE THREAD

**Issue by myitcv** [cue-lang/cue#1206]
### What version of CUE are you using (`cue version`)?

<pre>
$ cue version
cue version +0f53054d linux/amd64
</pre>

### Does this issue reproduce with the latest release?

Yes

### What did you do?

Follow up from https://github.com/cue-lang/cue/discussions/1203

Per https://github.com/cue-lang/cue/discussions/1203#discussioncomment-1188829:

```
# definition
exec cue eval def.cue
cmp stdout stdout.golden

# close
exec cue eval close.cue
cmp stdout stdout.golden

-- close.cue --
package y

parameters: close({
	name: string
	port: int

	{
		name: "custom-port"
	} | *{
		port: 8080
		name: "default-port"
	}

})

custom: parameters & {
	port: 9999
}
-- def.cue --
package y

#parameters: {
	name: string
	port: int

	{
		name: "custom-port"
	} | *{
		port: 8080
		name: "default-port"
	}

}

custom: #parameters & {
	port: 9999
}
-- stdout.golden --
#parameters: {
    name: "default-port"
    port: 8080
}
custom: {
    name: "custom-port"
    port: 9999
}
```

gives:

```
# definition (0.024s)
# close (0.018s)
> exec cue eval close.cue
[stderr]
custom.port: conflicting values 9999 and 8080:
    ./close.cue:10:9
    ./close.cue:16:9
    ./close.cue:17:8
[exit status 1]
FAIL: /tmp/testscript466980952/repro.txt/script.txt:6: unexpected command failure
```

From the error message it appears that the default of the disjunction is being selected "too early" for `close.cue`, which is what leads to the conflict. 

Quite possibly related however is the following:

```
exec cue eval -c

-- x.cue --
package a

x: close({
	{
		name: int
	} | {
		name: string
	}
})

y: x & {
	name: "Hello"
}
```

which gives:

```
> exec cue eval -c
[stderr]
x: unresolved disjunction {name:int} | {name:string} (type struct):
    ./x.cue:3:4
y: unresolved disjunction {name:int} | {name:string} (type struct):
    ./x.cue:3:4
[exit status 1]
FAIL: /tmp/testscript781051936/repro.txt/script.txt:1: unexpected command failure
```

The error for `x` is expected, but the error for `y` is not. This suggests that with `close()` the evaluator tries to eliminate elements from the disjunction "too early", i.e. prior to the unification with the concrete struct `{name: "Hello"}`. Again, using a definition instead of `close()` makes the problem disappear. 

### What did you expect to see?

As above.

### What did you see instead?

As above.

**Comment by mpvl** [comment]
Indeed, arguments to builtins are, as it stands, always evaluated. Just as this is the case for all binary operators but `|` and `&`, we could perhaps also make an exception for some builtins.

**Comment by mvdan** [comment]
Fixed at tip. The `close({...})` form now evaluates identically to the
`#parameters` definition form — `custom` resolves to
`{name: "custom-port", port: 9999}` with no `conflicting values 9999 and
8080`.

The second reproducer is fixed too: only `x` reports an incomplete
disjunction, as expected, while `y` resolves to `{name: "Hello"}`.

I will close this with a regression test. Thanks!
